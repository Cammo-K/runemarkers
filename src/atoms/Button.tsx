import { useEffect, useState } from "react";
import styles from "./Button.module.css";

export default function Button(
  buttonProps: React.ButtonHTMLAttributes<HTMLButtonElement>
) {
  const [isAnimating, setIsAnimating] = useState(false);
  useEffect(() => {
    const timeout = setTimeout(() => {
      setIsAnimating(false);
    }, 1100);
    return () => clearTimeout(timeout);
  }, [isAnimating]);

  return (
    <button
      className={styles.button}
      {...buttonProps}
      onClick={(e) => {
        if (buttonProps.onClick) {
          buttonProps.onClick(e);
        }
        setIsAnimating(true);
      }}
    >
      {buttonProps.children}
      <div
        className={[
          styles.animBar,
          isAnimating ? styles.animBarActive : "",
        ].join(" ")}
      ></div>
    </button>
  );
}
