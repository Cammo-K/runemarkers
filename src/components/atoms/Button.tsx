import { ButtonHTMLAttributes, useEffect, useState } from 'react';
import styles from '@/components/atoms/Button.module.css';

export default function Button(
  buttonProps: ButtonHTMLAttributes<HTMLButtonElement>,
) {
  const [isAnimating, setIsAnimating] = useState(false);
  useEffect(() => {
    if (!isAnimating) {
      return;
    }
    const timeout = setTimeout(() => {
      setIsAnimating(false);
    }, 1100);
    return () => clearTimeout(timeout);
  }, [isAnimating]);

  return (
    <button
      {...buttonProps}
      className={
        buttonProps.className
          ? [buttonProps.className, styles.button].join(` `)
          : styles.button
      }
      onClick={(e) => {
        if (buttonProps.onClick) {
          buttonProps.onClick(e);
        }
        setIsAnimating(true);
      }}
    >
      {buttonProps.children}
      <div
        className={[
          styles.animBar,
          isAnimating ? styles.animBarActive : ``,
        ].join(` `)}
      ></div>
    </button>
  );
}
