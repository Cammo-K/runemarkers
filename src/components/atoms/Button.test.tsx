import '@testing-library/jest-dom';
import { act, fireEvent, render } from '@testing-library/react';
import Button from './Button';

const callback = jest.fn();

const testingButton = (
  <Button
    onClick={() => {
      callback();
    }}
  >
    Test
  </Button>
);

const testingButtonWithClassName = (
  <Button
    onClick={() => {
      callback();
    }}
    className="test-class"
  >
    Test
  </Button>
);

describe(`Button`, () => {
  it(`should render correctly`, () => {
    const { container } = render(testingButton);
    expect(container).toMatchSnapshot();
  });

  it(`should have a functioning onClick handler`, () => {
    const { container } = render(testingButton);
    const button = container.querySelector(`button`);
    act(() => {
      if (!button) {
        throw new Error(`No button found`);
      }
      fireEvent.click(button);
    });
    expect(callback).toHaveBeenCalled();
  });

  it(`should be able to animate correctly`, () => {
    jest.useFakeTimers();
    jest.spyOn(global, `setTimeout`);
    const { container } = render(testingButton);
    const button = container.querySelector(`button`);
    act(() => {
      if (!button) {
        throw new Error(`No button found`);
      }
      fireEvent.click(button);
    });
    expect(container.querySelector(`.animBarActive`)).toBeTruthy();
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(container.querySelector(`.animBarActive`)).toBeFalsy();
    expect(setTimeout).toHaveBeenCalledTimes(1);
  });

  it(`should join class names correctly`, () => {
    const { container } = render(testingButtonWithClassName);
    expect(
      container.querySelector(`button`)?.classList[0] === `test-class`,
    ).toBeTruthy();
  });
});
