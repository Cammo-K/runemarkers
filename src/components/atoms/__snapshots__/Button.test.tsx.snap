// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<div>
  <button
    class="button"
  >
    Test
    <div
      class="animBar "
    />
  </button>
</div>
`;
